### PR TITLE
Add ernie-ctm

### PR DESCRIPTION
**Pull request by kinghuin**
### PR types
New features

### PR changes
Models

### Describe
1. add ernie-ctm model
2. add wordtag example
3. fix crf and vit decode

### CODE REVIEW

File: examples/information_extraction/wordtag/README.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+```bash
+python -u ./train.py \
+    --max_seq_len 128 \
+    --batch_size 32   \
+    --learning_rate 1e-4 \
+    --num_train_epochs 3 \
+    --logging_steps 1 \
+    --save_steps 1 \
+    --output_dir ./tmp/ \
+    --n_gpu 1 \
```
Comment (ZeyuChen) [inline examples/information_extraction/wordtag/README.md]: 统一下--device和n_gpu的用法，与zeyang确定

File: examples/information_extraction/wordtag/data/tags.txt [hunk 2]
```diff
@@ -0,0 +1,265 @@
+B-人物类_实体
```
Comment (ZeyuChen) [inline examples/information_extraction/wordtag/data/tags.txt]: 数据集是否可以链接外挂，不要在github上面直接托管，会增大git的空间，拖慢后续下载

File: examples/information_extraction/wordtag/eval.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    parser.add_argument("--batch_size",default=32,type=int,help="Batch size per GPU/CPU for training.", )
+    parser.add_argument("--n_gpu",default=1,type=int,help="number of gpus to use, 0 for cpu.")
+    # yapf: enable
+
+    args = parser.parse_args()
+    return args
+
+
+def do_eval(args):
+    paddle.set_device("gpu" if args.n_gpu else "cpu")
```
Comment (ZeyuChen) [inline examples/information_extraction/wordtag/eval.py]: 改为规定的用法

File: examples/information_extraction/wordtag/eval.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    for arg, value in sorted(vars(args).items()):
+        print('%s: %s' % (arg, value))
+    print('------------------------------------------------')
+
+
+if __name__ == "__main__":
+    args = parse_args()
+    print_arguments(args)
+    if args.n_gpu > 1:
+        paddle.distributed.spawn(do_eval, args=(args, ), nprocs=args.n_gpu)
```
Comment (ZeyuChen) [inline examples/information_extraction/wordtag/eval.py]: 多卡训练改为launch方式

File: examples/information_extraction/wordtag/train.py [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+from metric import SequenceAccuracy
+
+
+def parse_args():
+    parser = argparse.ArgumentParser()
+
+    # yapf: disable
+    parser.add_argument("--data_dir", default="./data", type=str, help="The input data dir, should contain [train/test].json and [train/test]_metrics.json")
+    parser.add_argument("--model_dir", default="ernie-ctm", type=str, help="The pre-trained model checkpoint dir.")
+    parser.add_argument("--output_dir",default="./outpout_dir",type=str, help="The output directory where the model predictions and checkpoints will be written.",)
```
Comment (ZeyuChen) [inline examples/information_extraction/wordtag/train.py]: 虽然没yapf，但逗号后空格这些还是要有基础语法保障

Comment (ZeyuChen) [reply examples/information_extraction/wordtag/train.py]: `,type=str` -> `, type=str`

File: paddlenlp/layers/crf.py [hunk 6]
```diff
@@ -106,13 +106,15 @@ def forward(self, inputs, lengths):
...
-                mat = input_exp
+                # mat = input_exp
+                alpha = inputs[:, 0]
             else:
                 alpha_exp = alpha.unsqueeze(1).expand(
                     [batch_size, n_labels, n_labels])
                 # F(n) = logsumexp(F(n-1) + p(y_n) + T(y_{n-1}, y_n))
                 mat = input_exp + trans_exp + alpha_exp
-            alpha = paddle.logsumexp(mat, 2)
+                # import pdb; pdb.set_trace()
```
Comment (ZeyuChen) [inline paddlenlp/layers/crf.py]: 不要引入pdb代码

File: examples/information_extraction/wordtag/README.md [hunk 7]
```diff
@@ -0,0 +1,55 @@
+# ERNIE-CTM
+
+## 简介
+
+ERNIE for **C**hinese **T**ext **M**ining，适用于中文挖掘任务的预训练语言模型，包含全中文词表、中英文标点、单位符号、汉语拼音等token，大幅减少tokenize中UNK的情况，对中文标注、挖掘类任务有较大增益。
```
Comment (ZeyuChen) [inline examples/information_extraction/wordtag/README.md:5]: 建议简介让KG同学优化下，感觉这段简介有点普通

File: paddlenlp/transformers/ernie_ctm/modeling.py [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+
+import paddle
+import paddle.nn as nn
+import paddle.tensor as tensor
+from paddle.nn import TransformerEncoder, Linear, Layer, Embedding, LayerNorm, Tanh
+from paddlenlp.layers.crf import LinearChainCrf, ViterbiDecoder, LinearChainCrfLoss
+
+from .. import PretrainedModel, register_base_model
+
+__all__ = ['ErnieCtmModel', 'ErnieCtmWordtagModel']
```
Comment (ZeyuChen) [inline paddlenlp/transformers/ernie_ctm/modeling.py:23]: 只能做这两个任务吗？不能做其他的Fineune任务？

**Review by ZeyuChen — COMMENTED** [review]
统一下启动方式，细化README

File: examples/information_extraction/wordtag/README.md (hunk 1 above)
Comment (wawltor) [reply examples/information_extraction/wordtag/README.md]: 这块也统一分布式的用法吧，可以提供一个python train.py单卡版本，同时也提供一个lanuch方式启动多卡版本吧

File: examples/information_extraction/wordtag/README.md [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+    --num_train_epochs 3 \
+    --logging_steps 1 \
+    --save_steps 1 \
+    --output_dir ./tmp/ \
+    --n_gpu 1 \
+```
+
+### 模型评估
+
+通过加载训练保存的模型，可以对测试集数据进行验证，启动方式如下：
```
Comment (wawltor) [inline examples/information_extraction/wordtag/README.md]: 有点不通顺，建议改成训练过程中保存的模型

File: examples/information_extraction/wordtag/README.md [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+
+```bash
+python -u ./eval.py \
+    --max_seq_len 128 \
+    --batch_size 32   \
+    --model_dir ./tmp/ernie_ctm_ft_model_1.pdparams \
+    --n_gpu 1 \
+```
```
Comment (wawltor) [inline examples/information_extraction/wordtag/README.md:77]: 同上

File: examples/information_extraction/wordtag/README.md [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+
+其中 model_dir 是模型加载路径。
```
Comment (wawltor) [inline examples/information_extraction/wordtag/README.md]: 这块最好能提示一下，要根据具体的模型路径填写该项

File: examples/information_extraction/wordtag/eval.py [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+
+import os
+import argparse
+import time
+import random
+from functools import partial
+
+import paddle
+from paddle.io import DataLoader
+import numpy as np
```
Comment (wawltor) [inline examples/information_extraction/wordtag/eval.py]: numpy 提前

File: examples/information_extraction/wordtag/eval.py [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+from data import load_dataset, load_dict, convert_example
+from metric import SequenceAccuracy
+
+
+def parse_args():
+    parser = argparse.ArgumentParser()
+
+    # yapf: disable
+    parser.add_argument("--data_dir", default="./data", type=str, help="The input data dir, should contain [train/test].json and [train/test]_metrics.json")
+    parser.add_argument("--model_dir", default="ernie-ctm", type=str, help="The pre-trained model checkpoint dir.")
```
Comment (wawltor) [inline examples/information_extraction/wordtag/eval.py]: 这里的model_dir是不是有点问题，我理解model_dir这里可以是预训练模型的名字或者路径

File: examples/information_extraction/wordtag/metric.py [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import paddle
+
+
+class SequenceAccuracy():
```
Comment (wawltor) [inline examples/information_extraction/wordtag/metric.py]: 这块应该要继承一下框架的paddle.metric.Metric基类

File: examples/information_extraction/wordtag/predict.py [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+
+import os
+import argparse
+import time
+import random
+from functools import partial
+
+import paddle
+from paddle.io import DataLoader
+import numpy as np
```
Comment (wawltor) [inline examples/information_extraction/wordtag/predict.py]: numpy提前

File: examples/information_extraction/wordtag/predict.py [hunk 16]
```diff
@@ -0,0 +1,167 @@
...
+from data import load_dataset, load_dict, convert_example
+from metric import SequenceAccuracy
+
+
+def parse_args():
+    parser = argparse.ArgumentParser()
+
+    # yapf: disable
+    parser.add_argument("--data_dir", default="./data", type=str, help="The input data dir, should contain [train/test].json and [train/test]_metrics.json")
+    parser.add_argument("--model_dir", default="ernie-ctm", type=str, help="The pre-trained model checkpoint dir.")
```
Comment (wawltor) [inline examples/information_extraction/wordtag/predict.py]: 同上

File: examples/information_extraction/wordtag/predict.py [hunk 17]
```diff
@@ -0,0 +1,167 @@
...
+    model = ErnieCtmWordtagModel.from_pretrained(
+        args.model_dir,
+        num_cls_label=len(labels_to_idx),
+        num_tag=len(tags_to_idx),
+        ignore_index=tags_to_idx["O"])
+
+    if paddle.distributed.get_world_size() > 1:
+        model = paddle.DataParallel(model)
+
+    model.eval()
```
Comment (wawltor) [inline examples/information_extraction/wordtag/predict.py]: 这里要paddle.no_grad

File: examples/information_extraction/wordtag/train.py [hunk 18]
```diff
@@ -0,0 +1,258 @@
...
+        num_cls_label=len(labels_to_idx),
+        num_tag=len(tags_to_idx),
+        ignore_index=tags_to_idx["O"])
+
+    if paddle.distributed.get_world_size() > 1:
+        model = paddle.DataParallel(model)
+
+    num_training_steps = args.max_steps if args.max_steps > 0 else (
+        len(train_data_loader) * args.num_train_epochs)
+    warmup = args.warmup_steps if args.warmup_steps > 0 else args.warmup_proportion
```
Comment (wawltor) [inline examples/information_extraction/wordtag/train.py:105]: 这里的warmuo是不是已经 args.warmup_proportion * num_train_steps

File: examples/information_extraction/wordtag/train.py [hunk 19]
```diff
@@ -0,0 +1,258 @@
...
+        train_ds) / args.batch_size * args.num_train_epochs
+
+    decay_params = [
+        p.name for n, p in model.named_parameters()
+        if not any(nd in n for nd in ["bias", "norm"])
+    ]
+    optimizer = paddle.optimizer.AdamW(
+        learning_rate=lr_scheduler,
+        beta1=0.9,
+        beta2=0.999,
```
Comment (wawltor) [inline examples/information_extraction/wordtag/train.py]: 这里的beta1 beta2是不是默认值就可以

File: examples/information_extraction/wordtag/eval.py [hunk 20]
```diff
@@ -12,30 +12,22 @@
...
 def parse_args():
     parser = argparse.ArgumentParser()
 
     # yapf: disable
     parser.add_argument("--data_dir", default="./data", type=str, help="The input data dir, should contain [train/test].json and [train/test]_metrics.json")
-    parser.add_argument("--model_dir", default="ernie-ctm", type=str, help="The pre-trained model checkpoint dir.")
+    parser.add_argument("--init_ckpt_dir", default="ernie-ctm", type=str, help="The pre-trained model checkpoint dir.")
     parser.add_argument("--max_seq_len",default=128,type=int,help="The maximum total input sequence length after tokenization. Sequences longer than this will be truncated, sequences shorter will be padded.", )
     parser.add_argument("--batch_size",default=32,type=int,help="Batch size per GPU/CPU for training.", )
     parser.add_argument("--n_gpu",default=1,type=int,help="number of gpus to use, 0 for cpu.")
```
Comment (ZeyuChen) [inline examples/information_extraction/wordtag/eval.py]: 全部统一为--device的argparse + choices方案来控制device

File: examples/information_extraction/wordtag/eval.py (hunk 3 above)
Comment (ZeyuChen) [reply examples/information_extraction/wordtag/eval.py]: 改为--device

File: examples/information_extraction/wordtag/metric.py [hunk 21]
```diff
@@ -47,3 +48,177 @@ def accumulate(self):
...
+            a_l, b_l = a_r + 1, b_r + 1
+            continue
+        p += 1
+    return acc
+
+
+def wordseg_soft_acc(list_a: List[Tuple[str, str]],
+                     list_b: List[Tuple[str, str]]) -> float:
+    """
+    Calculate extra metrics of word-seg
```
Comment (ZeyuChen) [inline examples/information_extraction/wordtag/metric.py:171]: 不太理解这段代码的作用，包括wordseg这里是否指的是内部的分词工具？

**Review by ZeyuChen — COMMENTED** [review]
Unifiy the device luanch and add comments for wordtag and wordseg acc

File: examples/information_extraction/wordtag/README.md (hunk 1 above)
Comment (kinghuin) [reply examples/information_extraction/wordtag/README.md]: DONE

File: examples/information_extraction/wordtag/data/tags.txt (hunk 2 above)
Comment (kinghuin) [reply examples/information_extraction/wordtag/data/tags.txt]: DONE

File: examples/information_extraction/wordtag/eval.py (hunk 3 above)
Comment (kinghuin) [reply examples/information_extraction/wordtag/eval.py]: DONE

File: examples/information_extraction/wordtag/eval.py (hunk 4 above)
Comment (kinghuin) [reply examples/information_extraction/wordtag/eval.py]: DONE

File: examples/information_extraction/wordtag/train.py (hunk 5 above)
Comment (kinghuin) [reply examples/information_extraction/wordtag/train.py]: DONE

File: paddlenlp/layers/crf.py (hunk 6 above)
Comment (kinghuin) [reply paddlenlp/layers/crf.py]: DONE

File: examples/information_extraction/wordtag/README.md (hunk 7 above)
Comment (kinghuin) [reply examples/information_extraction/wordtag/README.md:5]: Waiting

File: paddlenlp/transformers/ernie_ctm/modeling.py (hunk 8 above)
Comment (kinghuin) [reply paddlenlp/transformers/ernie_ctm/modeling.py:23]: 预训练模型中包含了crf参数，特用于Wordtag任务

File: examples/information_extraction/wordtag/README.md (hunk 9 above)
Comment (kinghuin) [reply examples/information_extraction/wordtag/README.md]: DONE

File: examples/information_extraction/wordtag/README.md (hunk 10 above)
Comment (kinghuin) [reply examples/information_extraction/wordtag/README.md:77]: DONE

File: examples/information_extraction/wordtag/README.md (hunk 11 above)
Comment (kinghuin) [reply examples/information_extraction/wordtag/README.md]: DONE

File: examples/information_extraction/wordtag/eval.py (hunk 12 above)
Comment (kinghuin) [reply examples/information_extraction/wordtag/eval.py]: Deleted

File: examples/information_extraction/wordtag/eval.py (hunk 13 above)
Comment (kinghuin) [reply examples/information_extraction/wordtag/eval.py]: 改为 init_ckpt_path

File: examples/information_extraction/wordtag/metric.py (hunk 14 above)
Comment (kinghuin) [reply examples/information_extraction/wordtag/metric.py]: DONE

File: examples/information_extraction/wordtag/predict.py (hunk 15 above)
Comment (kinghuin) [reply examples/information_extraction/wordtag/predict.py]: Deleted

File: examples/information_extraction/wordtag/predict.py (hunk 16 above)
Comment (kinghuin) [reply examples/information_extraction/wordtag/predict.py]: DONE

File: examples/information_extraction/wordtag/predict.py (hunk 17 above)
Comment (kinghuin) [reply examples/information_extraction/wordtag/predict.py]: Deleted

File: examples/information_extraction/wordtag/train.py (hunk 18 above)
Comment (kinghuin) [reply examples/information_extraction/wordtag/train.py:105]: 是的

File: examples/information_extraction/wordtag/train.py (hunk 19 above)
Comment (kinghuin) [reply examples/information_extraction/wordtag/train.py]: Deleted

File: examples/information_extraction/wordtag/eval.py (hunk 20 above)
Comment (kinghuin) [reply examples/information_extraction/wordtag/eval.py]: DONE

**Review by wawltor — DISMISSED** [review]
LGTM

**Review by wawltor — APPROVED** [review]
LGTM
